src/elements/segment.rs

```diff
@@ -72,49 +99,67 @@ impl Serialize for ElementSegment {
 #[derive(Clone, Debug, PartialEq)]
 pub struct DataSegment {
 	index: u32,
-	offset: InitExpr,
+	offset: Option<InitExpr>,
 	value: Vec<u8>,
+	passive: bool,
 }
 
 impl DataSegment {
 	/// New data segments.
-	pub fn new(index: u32, offset: InitExpr, value: Vec<u8>) -> Self {
+	pub fn new(index: u32, offset: Option<InitExpr>, value: Vec<u8>, passive: bool) -> Self {
 		DataSegment {
 			index: index,
 			offset: offset,
 			value: value,
+			passive: passive,
 		}
 	}
 
 	/// Linear memory index (currently the only valid value is `0`).
 	pub fn index(&self) -> u32 { self.index }
 
 	/// An i32 initializer expression that computes the offset at which to place the data.
-	pub fn offset(&self) -> &InitExpr { &self.offset }
+	pub fn offset(&self) -> &Option<InitExpr> { &self.offset }
 
 	/// An i32 initializer expression that computes the offset at which to place the data (mutable)
-	pub fn offset_mut(&mut self) -> &mut InitExpr { &mut self.offset }
+	pub fn offset_mut(&mut self) -> &mut Option<InitExpr> { &mut self.offset }
 
 	/// Initial value of the data segment.
 	pub fn value(&self) -> &[u8] { &self.value }
 
 	/// Initial value of the data segment (mutable).
 	pub fn value_mut(&mut self) -> &mut Vec<u8> { &mut self.value }
+
+	/// Whether or not this data segment is "passive"
+	pub fn passive(&self) -> bool { self.passive }
+
+	/// Whether or not this data segment is "passive" (mutable)
+	pub fn passive_mut(&mut self) -> &mut bool { &mut self.passive }
 }
 
 impl Deserialize for DataSegment {
 	 type Error = Error;
 
 	fn deserialize<R: io::Read>(reader: &mut R) -> Result<Self, Self::Error> {
-		let index = VarUint32::deserialize(reader)?;
-		let offset = InitExpr::deserialize(reader)?;
+		let flags: u8 = VarUint7::deserialize(reader)?.into();
```

Comment:
This one is really tough.

The proposed spec states that it should be a `uint8` rather than `varuint7`. But I don't understand this completely and how is that a backward compatible change: 

First, In the MVP `index` is represented by `varuint32`. Second, the spec allows for non-cannonical encodings, so `0:u32` can be encoded as (in hex) `00`, `80 00`, `80 80 00`, `80 80 80 00` and maybe `80 80 80 80 00`, though I'm not sure about the last one.

`uint8` supports the whole range [00, FF], but it's a single byte.

On the other hand, `VarUint7` supports values only in range `[0x00, 0x7F]`, but because it's a `varuint7` it is an error if it encounters `0x80` AFAIK (the error is unimplemented for now though), so it is essentially a single byte as well.

Reply:
I've opened https://github.com/WebAssembly/bulk-memory-operations/issues/34 to track this, would probably be good to move discussion there!

Reply:
Thank you very much, I really appreciate it!